### ISSUE TITLE
sh:class in PropertyShape fails to validate after 2+ subClassOf hops

### ISSUE THREAD

**Issue by gtfierro** [RDFLib/pySHACL#87]
Hello!

First of all, I want to thank you all for the wonderful work you have been doing on pySHACL. It is a critical component of my day-to-day work and it has been extraordinarily stable and easy to work with.

I'm running into what I believe is a bug with `sh:class` validation. I'm running 

```
rdflib==6.0.0
pyshacl==0.15.0
```

Here's a minimum reproducing example:

```python
import pyshacl
import rdflib

ttl_file = """
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix brick: <https://brickschema.org/schema/Brick#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix : <urn:ex#> .

brick:VAV a owl:Class .

brick:Point a owl:Class .

brick:Sensor a owl:Class ;
    rdfs:subClassOf brick:Point .

brick:Temperature_Sensor a owl:Class ;
    rdfs:subClassOf brick:Sensor .

brick:Air_Temperature_Sensor a owl:Class ;
    rdfs:subClassOf brick:Temperature_Sensor .

brick:hasPoint  a   owl:ObjectProperty .

:shape a sh:NodeShape ;
    sh:targetClass brick:VAV ;
    sh:property :haspointshape .

:haspointshape a sh:PropertyShape ;
    sh:path brick:hasPoint ;
    sh:class brick:Point .

:s0 a brick:Sensor .
:vav0 a brick:VAV ;
    brick:hasPoint :s0 .

:s1 a brick:Sensor .
:vav1 a brick:VAV ;
    brick:hasPoint :s1 .

:s2 a brick:Temperature_Sensor .
:vav2 a brick:VAV ;
    brick:hasPoint :s2 .

:s3 a brick:Air_Temperature_Sensor .
:vav3 a brick:VAV ;
    brick:hasPoint :s3 .
"""

g = rdflib.Graph()
g.parse(data=ttl_file, format="ttl")

valid, _, report = pyshacl.validate(g)
assert valid, report
```

I get the following report:

```
AssertionError: Validation Report
Conforms: False
Results (2):
Constraint Violation in ClassConstraintComponent (http://www.w3.org/ns/shacl#ClassConstraintComponent):
        Severity: sh:Violation
        Source Shape: :haspointshape
        Focus Node: :vav2
        Value Node: :s2
        Result Path: brick:hasPoint
        Message: Value does not have class brick:Point
Constraint Violation in ClassConstraintComponent (http://www.w3.org/ns/shacl#ClassConstraintComponent):
        Severity: sh:Violation
        Source Shape: :haspointshape
        Focus Node: :vav3
        Value Node: :s3
        Result Path: brick:hasPoint
        Message: Value does not have class brick:Point
```

`:s1`'s type is 1 `rdfs:subClassOf` hop away from the object of `sh:class` (`brick:Point`) and validates fine. `:s2`'s type is 2 `rdfs:subClassOf` hops away from `brick:Point` and fails to validate. `:s3`'s type is 3 hops away and similarly fails. As per [the SHACL definition](https://www.w3.org/TR/shacl/#ClassConstraintComponent) I'd expect this all of the `:s{0,1,2,3}` instances to validate.

Can you verify that you can reproduce the above issue, or let me know if I've made a mistake? I haven't had the time to start looking into the pySHACL implementation yet to see if I can determine where the issue lies, but I plan to spend some time doing so.

**Comment by gtfierro** [comment]
I just found [this annotation](https://github.com/RDFLib/pySHACL/blob/master/pyshacl/constraints/core/value_constraints.py#L116-L122) in the code (reproduced below):

> Note, this only ones _one_ level of subclass traversing.
> For more levels, the whole target graph should be put through
> a RDFS reasoning engine.

We would like to avoid having to apply the RDFS reasoning engine for our use case because it pollutes the graph with the materialized `<instance> rdf:type <superclass>` triples which makes it difficult to determine via SPARQL queries what the most *specific* or immediate type of an entity is. It is also not technically to spec with the [SHACL definition of `sh:class`](https://www.w3.org/TR/shacl/#ClassConstraintComponent).

Perhaps an "easy" fix would be the use of `target_graph.transitive_objects` rather than `target_graph.objects` on [this line](https://github.com/RDFLib/pySHACL/blob/master/pyshacl/constraints/core/value_constraints.py#L119)?

**Comment by ashleysommer** [comment]
Hi @gtfierro 
Thanks for the well written and detailed issue and PR.

As you have seen (from the comment in the code) it was a deliberate decision of PySHACL to include just a single level of `rdf:subClassOf` checking. That was because it was able to be quickly and easily implemented with a single check of `rdf:type` and `rdfs:subClassOf`, without the need for recursion or transitive queries.

I think I also had made an incorrect interpretation of the `ClassConstraintComponent` details in the SHACL Spec.
>For each value node that is either a literal, or a non-literal that is not a SHACL instance of **$class** in the data graph, there is a validation result with the value node as sh:value. 

When implementing `ClassConstraintComponent` I interpreted that as the RDF definition of "instance of". That is, check only for immediate `rdf:type`. I think I added the single `rdfs:subClassOf` check as a convenience, because one early beta tested needed it, and a single extra `subClassOf` query didn't affect the performance of the validator.

However after reading your issue, I double checked. This is the part I missed:
>**SHACL Class Instance**
A node n in an RDF graph **G** is a SHACL instance of a SHACL class **C** in **G** if one of the SHACL types of _n_ in **G** is **C**.

>**SHACL Type**
The SHACL types of an RDF term in an RDF graph is the set of its values for `rdf:type` in the graph as well as the SHACL superclasses of these values in the graph.

>**SHACL Subclass, SHACL superclass**
A node Sub in an RDF graph is a SHACL subclass of another node Super in the graph if there is a sequence of triples in the graph each with predicate `rdfs:subClassOf` such that the subject of the first triple is Sub, the object of the last triple is Super, and the object of each triple except the last is the subject of the next. If Sub is a SHACL subclass of Super in an RDF graph then Super is a SHACL _superclass_ of Sub in the graph.

That is a very sterile way of defining it, but it looks like you're right, and the `ClassConstraintComponent` should check transitively for `subClassOf` relationships. I'll check your PR, merge it in.
